src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -836,27 +888,730 @@ static MemorySegment mapFile(Path path, long bytesOffset, long bytesSize, FileCh
+     * Supported array types are {@code byte[]}, {@code char[]},{@code short[]},{@code int[]},{@code float[]},{@code long[]} and {@code double[]}.
+     */
+    @ForceInline
+    static void copy(
+            MemorySegment srcSegment, ValueLayout srcLayout, long srcOffset,
+            Object dstArray, int dstIndex, int elementCount) {
+        Objects.requireNonNull(srcSegment);
+        Objects.requireNonNull(dstArray);
+        Objects.requireNonNull(srcLayout);
+        long baseAndScale = getBaseAndScale(dstArray.getClass());
+        int dstBase = (int)baseAndScale;
+        int dstWidth = (int)(baseAndScale >> 32);
+        AbstractMemorySegmentImpl srcImpl = (AbstractMemorySegmentImpl)srcSegment;
+        srcImpl.checkAccess(srcOffset, elementCount * dstWidth, true);
+        Objects.checkFromIndexSize(dstIndex, elementCount, Array.getLength(dstArray));
+        if (dstWidth == 1 || srcLayout.order() == ByteOrder.nativeOrder()) {
+            ScopedMemoryAccess.getScopedMemoryAccess().copyMemory(srcImpl.scope(), null,
+                    srcImpl.unsafeGetBase(), srcImpl.unsafeGetOffset() + srcOffset,
+                    dstArray, dstBase + (dstIndex * dstWidth), elementCount * dstWidth);
+        } else {
+            ScopedMemoryAccess.getScopedMemoryAccess().copySwapMemory(srcImpl.scope(), null,
+                    srcImpl.unsafeGetBase(), srcImpl.unsafeGetOffset() + srcOffset,
+                    dstArray, dstBase + (dstIndex * dstWidth), elementCount * dstWidth, dstWidth);
+        }
+    }
+
+    /**
+     * Copies a number of elements from a source array to a destination segment,
+     * starting at a given array index, and a given segment offset (expressed in bytes), using the given destination element layout.
+     * @param srcArray the source array.
+     * @param srcIndex the starting index of the source array.
+     * @param dstSegment the destination segment.
+     * @param dstLayout the destination element layout. If the byte order associated with the layout is
+     * different from the native order, a byte swap operation will be performed on each array element.
+     * @param dstOffset the starting offset, in bytes, of the destination segment.
+     * @param elementCount the number of array elements to be copied.
+     * @throws  IllegalArgumentException if {@code srcArray} is not an array, or if it is an array but whose type is not supported.
+     * Supported array types are {@code byte[]}, {@code char[]},{@code short[]},{@code int[]},{@code float[]},{@code long[]} and {@code double[]}.
+     */
+    @ForceInline
+    static void copy(
+            Object srcArray, int srcIndex,
+            MemorySegment dstSegment, ValueLayout dstLayout, long dstOffset, int elementCount) {
+        Objects.requireNonNull(srcArray);
+        Objects.requireNonNull(dstSegment);
+        Objects.requireNonNull(dstLayout);
```

Comment:
same here: dstLayout's carrier type should be srcArray's type.

Reply:
Good point(s) - I will fix.